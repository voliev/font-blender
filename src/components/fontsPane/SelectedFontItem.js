import React, { Component } from "react";
import RemoveIcon from "@material-ui/icons/RemoveCircleOutlineRounded";
import PropTypes from "prop-types";
import "./styles/SelectedFontItem.css";

class SelectedFontItem extends Component {
  constructor() {
    super();

    this.item = React.createRef();

    this.handleButtonClick = this.handleButtonClick.bind(this);
    this.handleTransitionEnd = this.handleTransitionEnd.bind(this);
  }

  handleButtonClick(e) {
    const { target } = e;

    if (target.closest("button")) {
      this.item.current.classList.remove("active");
    }
  }

  handleTransitionEnd(e) {
    e.persist();

    if (e.propertyName === "transform") {
      return;
    }

    const { target } = e;
    const { family, fontDeselectionHandler } = this.props;

    if (!target.classList.contains("active")) {
      fontDeselectionHandler(family);
    }
  }

  componentDidMount() {
    if (this.item.current) {
      this.item.current.classList.add("active");
    }
  }

  render() {
    const { family } = this.props;

    return (
      <li
        className="SelectedFontItem"
        onTransitionEnd={this.handleTransitionEnd}
        ref={this.item}
      >
        <p className="font-family">{family}</p>
        <button
          className="btn btn-round-small"
          onClick={this.handleButtonClick}
          title="Deselect font"
          value={family}
        >
          <RemoveIcon fontSize="default" />
        </button>
      </li>
    );
  }
}

SelectedFontItem.propTypes = {
  family: PropTypes.string.isRequired,
  fontDeselectionHandler: PropTypes.func.isRequired
};

export default SelectedFontItem;
