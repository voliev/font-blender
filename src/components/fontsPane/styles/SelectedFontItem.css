.SelectedFontItem {
  align-items: center;
  display: flex;
  font-weight: 700;
  justify-content: space-between;
  height: 0;
  opacity: 0;
  overflow: hidden;
  transition: height 0.2s, opacity 0.2s;

}

.SelectedFontItem.active {
  height: 2.75rem;
  opacity: 1;
  transition: height 0.2s, opacity 0.2s;
}

.SelectedFontItem:last-child {
  margin: 0;
}

.SelectedFontItem .font-family {
  margin: 0 0 0 var(--space-lg);
}

.SelectedFontItem .btn {
  color: var(--red);
  margin: 0 var(--space-lg) 0 0;
  transition: color 0.1s ease-out, transform 0.1s ease-out;
}

.SelectedFontItem .btn:hover {
  transform: scale(1.2);
}
