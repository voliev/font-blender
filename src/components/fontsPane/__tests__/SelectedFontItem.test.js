import React from "react";
import { mount, shallow } from "enzyme";
import SelectedFontItem from "../SelectedFontItem";

const setup = (renderFn = shallow, propsOverrides = {}) => {
  const props = {
    family: "Roboto",
    fontDeselectionHandler: jest.fn(),
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: renderFn(<SelectedFontItem {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const { family, fontDeselectionHandler, wrapper } = setup(mount);

  expect(typeof wrapper.prop("family")).toBe("string");
  expect(wrapper.prop("family")).toBe(family);

  expect(typeof wrapper.prop("fontDeselectionHandler")).toBe("function");
  expect(wrapper.prop("fontDeselectionHandler")).toBe(fontDeselectionHandler);
});

it("should add class 'active'", () => {
  const componentDidMount = jest.spyOn(
    SelectedFontItem.prototype,
    "componentDidMount"
  );
  const { wrapper } = setup(mount);
  const instance = wrapper.instance();

  expect(componentDidMount).toHaveBeenCalledTimes(1);
  expect(instance.item.current.classList.contains("active")).toBeTruthy();
});

it("should handle button click", () => {
  const handleButtonClick = jest.spyOn(
    SelectedFontItem.prototype,
    "handleButtonClick"
  );
  const { wrapper } = setup(mount);
  const instance = wrapper.instance();

  expect(instance.item.current.classList.contains("active")).toBeTruthy();
  wrapper.find("button").simulate("click", {
    target: {
      closest: () => false
    }
  });

  expect(handleButtonClick).toHaveBeenCalledTimes(1);
  expect(instance.item.current.classList.contains("active")).toBeTruthy();

  wrapper.find("button").simulate("click", {
    target: {
      closest: () => true
    }
  });

  expect(handleButtonClick).toHaveBeenCalledTimes(2);
  expect(instance.item.current.classList.contains("active")).toBeFalsy();
});

it("should properly handle transition end", () => {
  const handleTransitionEnd = jest.spyOn(
    SelectedFontItem.prototype,
    "handleTransitionEnd"
  );
  const { family, fontDeselectionHandler, wrapper } = setup(mount);

  wrapper.find("li").simulate("transitionend", {
    target: {
      classList: {
        contains: () => false
      }
    }
  });

  expect(handleTransitionEnd).toHaveBeenCalledTimes(1);
  expect(fontDeselectionHandler).toHaveBeenCalledTimes(1);
  expect(fontDeselectionHandler).toHaveBeenLastCalledWith(family);

  wrapper.find("li").simulate("transitionend", {
    target: {
      classList: {
        contains: () => true
      }
    }
  });

  expect(handleTransitionEnd).toHaveBeenCalledTimes(2);
  expect(fontDeselectionHandler).toHaveBeenCalledTimes(1);
});

it("should ignore transform transition end", () => {
  const { wrapper, fontDeselectionHandler } = setup(mount);

  wrapper.instance().handleTransitionEnd({
    persist: () => {},
    propertyName: "transform"
  });

  expect(fontDeselectionHandler).toHaveBeenCalledTimes(0);
});
