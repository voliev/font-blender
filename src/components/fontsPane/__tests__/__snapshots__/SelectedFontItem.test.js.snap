// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render without crashing 1`] = `
<li
  className="SelectedFontItem"
  onTransitionEnd={[Function]}
>
  <p
    className="font-family"
  >
    Roboto
  </p>
  <button
    className="btn btn-round-small"
    onClick={[Function]}
    title="Deselect font"
    value="Roboto"
  >
    <RemoveCircleOutlineRoundedIcon
      fontSize="default"
    />
  </button>
</li>
`;
